src/prefect/orion/services/scheduler.py

```diff
@@ -116,14 +117,47 @@ async def run_once(self, db: OrionDBInterface):
     @inject_db
     def _get_select_deployments_to_schedule_query(self, db: OrionDBInterface):
         """
-        Returns a sqlalchemy query for selecting deployments to schedule
+        Returns a sqlalchemy query for selecting deployments to schedule.
+
+        The query gets the IDs of any deployments with:
+            - an active schedule
+            - EITHER:
+                - fewer than `min_runs` auto-scheduled runs
+                - OR the max scheduled time is less than `max_scheduled_time` in the future
         """
+        now = pendulum.now("UTC")
         query = (
             sa.select(db.Deployment.id)
+            .select_from(db.Deployment)
+            # TODO: on Postgres, this could be replaced with a lateral join that
+            # sorts by `next_scheduled_start_time desc` and limits by
+            # `self.min_runs` for a ~ 50% speedup. At the time of writing,
+            # performance of this universal query appears to be fast enough that
+            # this optimization is not worth maintaining db-specific queries
+            .join(
+                db.FlowRun,
+                # join on matching deployments, only picking up future scheduled runs
+                sa.and_(
+                    db.Deployment.id == db.FlowRun.id,
+                    db.FlowRun.state_type == StateType.SCHEDULED,
```

Comment:
What happens if I move a scheduled run to a "CANCELLED" state. Will this schedule it again on the next scheduler iteration?

Reply:
Good question. I think that would fall back into something similar to the current behavior where the scheduler would see, for example, that there are only two scheduled runs instead of three and seek to create the "missing" run. However the run wouldn't actually be created because of the idempotency key. 

In the scheme of things, while this would represent a temporary performance regression for that deployment, it does recover the correct behavior and I don't think we need to worry about it.

Reply:
@madkinsz i added a test for this situation